### PR TITLE
Add debuginfod support

### LINKED ISSUE #734: Add debuginfod support to dyninst
Debuginfod is a lightweight web service that indexes ELF/DWARF debugging resources by build-id and serves them over HTTP. Debuginfod support enables dyninst to query debuginfod servers for a file's separate debuginfo when it otherwise cannot be found. Debuginfod is distributed with elfutils, for more information see https://sourceware.org/elfutils/Debuginfod.html

[debuginfod.diff](https://github.com/dyninst/dyninst/files/4441085/debuginfod.txt)
This patch adds debuginfod support to dyninst. It calls the debuginfod client library function for fetching separate debuginfo within Elf_X::findDebugFile(). It also adds a cmake option ENABLE_DEBUGINFOD to control whether dyninst is built with debuginfod support. This requires either having the debuginfod client library libdebuginfod and its header file installed or disabling sterile builds (debuginfod will be downloaded along with the rest of elfutils). 

Currently the patch does not include any tests for debuginfod support since it was not clear to me how to best integrate it into the testsuite. A typical test would involve stripping the mutatee's debuginfo, storing it in a location that dyninst cannot find it on its own, initializing a server that indexes the debuginfo and then running the mutatee. Should the test driver contain special handling for stripping the mutatee's debuginfo and initializing a server? Or maybe this should be carried out in a separate executable within the testsuite. Any suggestions are appreciated.

### PR DESCRIPTION
**Pull request by aaronmerey**
Debuginfod is a lightweight web service that indexes ELF/DWARF debugging
resources by build-id and serves them over HTTP.

This patch enables dyninst to query debuginfod servers for a file's
separate debuginfo when it otherwise cannot be found.

This patch also adds a cmake option -DENABLE_DEBUGINFOD to control
whether dyninst is built with debuginfod support.

This requires having the debuginfod client library (libdebuginfod)
and header installed.

Debuginfod is distributed with elfutils, for more information see
https://sourceware.org/elfutils/Debuginfod.html

Closes #734

### CODE REVIEW

**Comment by aaronmerey** [comment]
See issue #734

**Comment by aaronmerey** [comment]
Ping, just wondering if there's been any progress setting up a local debuginfod server to verify the patch. debuginfod should be built from elfutils 0.179. Just include the configure option '--enable-debuginfod'. This requires having libmicrohttpd, libcurl, sqlite3 and libarchive installed.

Once installed, set the DEBUGINFOD_URLS environment variable to 'http://localhost:8002' (the server listens to port 8002 by default). A debuginfod server will start up and index a mutatee's separate debuginfo with the following command:

$ debuginfod -F <directory containing separate debuginfo>

If this debuginfo file is stored in a location that Elf_X::findDebugFile() cannot normally locate and dyninst is built with the -DENABLE_DEBUGINFOD option added in this patch, then the file should be acquired via debuginfod_find_debuginfo() within Elf_X::findDebugFile().

**Comment by hainest** [comment]
There are only three Dyninst developers, and we're all currently fixing bugs to try to push the 10.2 release out. We greatly appreciate folks contributing code, but we just don't have the cycles to test this right now. I will try to take a look at it next week sometime, but I can't make any promises right now. I'm assuming this PR is working for you in your local environment?

**Comment by aaronmerey** [comment]
Not a problem. The PR is working for me locally.

File: cmake/ElfUtils.cmake [hunk 1]
```diff
@@ -28,7 +28,7 @@
 #
 #======================================================================================
 
-if(LibElf_FOUND AND LibDwarf_FOUND)
+if(LibElf_FOUND AND LibDwarf_FOUND AND (LibDebuginfod_FOUND OR NOT ENABLE_DEBUGINFOD))
```
[the remark this answers is not in the document]
Comment (aaronmerey) [reply cmake/ElfUtils.cmake:31]: Sure, will remove.

File: cmake/ElfUtils.cmake [hunk 2]
```diff
@@ -37,7 +37,12 @@ if(NOT UNIX)
 endif()
 
 # Minimum acceptable version of elfutils
-set(_min_version 0.178)
+if(ENABLE_DEBUGINFOD)
+  set(_min_version 0.179)
```
[the remark this answers is not in the document]
Comment (aaronmerey) [reply cmake/ElfUtils.cmake:41]: That's right, 0.178 should be fine here. 0.179 adds debuginfod_{get,set}_user_data functions but they are not used in this patch.

File: cmake/ElfUtils.cmake [hunk 3]
```diff
@@ -72,18 +77,30 @@ endforeach()
 
 find_package(LibElf ${ElfUtils_MIN_VERSION})
 
-# Don't search for libdw if we didn't find a suitable libelf
+# Don't search for libdw or libdebuginfod if we didn't find a suitable libelf
 if(LibElf_FOUND)
   find_package(LibDwarf ${ElfUtils_MIN_VERSION})
+  if (ENABLE_DEBUGINFOD)
+    find_package(LibDebuginfod ${ElfUtils_MIN_VERSION})
```
[the remark this answers is not in the document]
Comment (aaronmerey) [reply cmake/ElfUtils.cmake:84]: Ok will add.

File: elf/src/Elf_X.C [hunk 4]
```diff
@@ -1722,37 +1725,79 @@ bool Elf_X::findDebugFile(std::string origfilename, string &output_name, char* &
...
+
+     char *filename;
+     int fd = debuginfod_find_debuginfo(client,
+                                        (const unsigned char *)buildid.c_str(),
+                                        0, &filename);
+     debuginfod_end(client);
+
+     if (fd >= 0) {
+        string fname = string(filename);
+        free(filename);
```
[the remark this answers is not in the document]
Comment (aaronmerey) [reply elf/src/Elf_X.C:1785]: That's correct.

File: cmake/ElfUtils.cmake [hunk 5]
```diff
@@ -72,18 +77,30 @@ endforeach()
...
+    set(_eu_lib_dirs ${LibElf_LIBRARY_DIRS} ${LibDwarf_LIBRARY_DIRS} ${LibDebuginfod_LIBRARY_DIRS})
+    set(_eu_libs ${LibElf_LIBRARIES} ${LibDwarf_LIBRARIES} ${LibDebuginfod_LIBRARIES})
+  else()
+    set(_eu_root ${ElfUtils_ROOT_DIR})
+    set(_eu_inc_dirs ${LibElf_INCLUDE_DIRS} ${LibDwarf_INCLUDE_DIRS})
+    set(_eu_lib_dirs ${LibElf_LIBRARY_DIRS} ${LibDwarf_LIBRARY_DIRS})
+    set(_eu_libs ${LibElf_LIBRARIES} ${LibDwarf_LIBRARIES})
+  endif()
   add_library(ElfUtils SHARED IMPORTED)
+elseif(ENABLE_DEBUGINFOD AND NOT LibDebuginfod_FOUND)
```
[the remark this answers is not in the document]
Comment (aaronmerey) [reply cmake/ElfUtils.cmake:102]: Ack.

**Comment by hainest** [comment]
https://bottle.cs.wisc.edu/search?dyninst_branch=PR736&status=OK

LGTM. I ran it with and without debuginfod. We don't have any explicit test for that, but it didn't introduce any regressions. If things are working on your end, I'll merge.

**Comment by aaronmerey** [comment]
Great. Things are working on my end too.

**Pull request merged**